File: js/3.js
const fs = require('fs');
const path = require("path");

const input = fs
	.readFileSync(path.resolve(__dirname, '../inputs/input3.txt'), 'utf8')
	.toString()
	.trim()
	.split('\n')


const OXYGEN = 'oxygen'
const CO2 = 'co2'


const preCount = (lines) => {
	const size_line = lines[0].trim().length
	const size_input = lines.length
	const counts = {}
	for (let i = 0; i < size_line; i += 1) {
		counts[i] = 0
	}

	for (const line of lines) {
		for (let i = 0; i < size_input; i += 1) {
			if (line[i] == '1') {
				counts[i] += 1
			}
		}
	}

	return counts
}


const getPowerConsumption = (lines) => {
	const counts = preCount(lines)
	const size = lines.length
	gamma = Object.values(counts).map((x) => x >= size / 2 ? 1 : 0)
								 .join('')

	epsilon = Object.values(counts).map((x) => x < size / 2 ? 1 : 0)
								   .join('')

	return parseInt(gamma, 2) * parseInt(epsilon, 2)
}



const getLifeSupportRating = (lines, mode, pos = 0) => {

	const size_input = lines.length
	let countOfOne = 0
	const lines_ones = []
	const lines_zeros = []

	for (const line of lines) {
		if (line[pos] === '1') {
			countOfOne += 1
			lines_ones.push(line)
		}
		else {
			lines_zeros.push(line)
		}			
	}
	
	const moreOnes = countOfOne >= size_input / 2

	let new_lines = []
	if ((moreOnes && mode === OXYGEN) || (!moreOnes && mode === CO2)) {
		new_lines = lines_ones
	}
	else {
		new_lines = lines_zeros
	}

	if (new_lines.length === 1) {
		return parseInt(new_lines[0], 2)
	}
		
	return getLifeSupportRating(new_lines, mode, pos + 1)
}


oxy = getLifeSupportRating(input, mode=OXYGEN)
co2 = getLifeSupportRating(input, mode=CO2)

console.log(`Result 1: ${getPowerConsumption(input)}`)
console.log(`Result 2: ${oxy * co2}`)

File: py/3.py

OXYGEN = 'oxygen'
CO2 = 'co2'


def preCount(lines):
	size_line = len(lines[0].strip())
	size_input = len(lines)
	counts = {i: 0 for i in range(size_line)}

	for line in lines:
		for i in range(size_line):
			counts[i] += 1 if line[i] == '1' else 0
	return counts


def getInt(array):
	binNum = ''.join([str(int(x)) for x in array])
	return int(binNum, 2)


def getPowerConsumption(lines):
	counts = preCount(lines)
	size = len(lines)
	gamma = getInt(map(lambda x:  x >= size / 2, counts.values()))
	epsilon = getInt(map(lambda x: x < size / 2, counts.values()))
	return gamma * epsilon


def getLifeSupportRating(lines, mode, pos = 0):

	size_input = len(lines)
	countOfOne = 0
	lines_ones, lines_zeros = [], []

	for line in lines:
		if line[pos] == '1':
			countOfOne += 1
			lines_ones.append(line)
		else:
			lines_zeros.append(line)

	moreOnes = countOfOne >= size_input / 2

	if (moreOnes and mode == OXYGEN) or (not moreOnes and mode == CO2):
		new_lines = lines_ones
	else:
		new_lines = lines_zeros

	if len(new_lines) == 1:
		return int(new_lines[0], 2)

	return getLifeSupportRating(new_lines, mode, pos + 1)


if __name__ == "__main__": 

	with open('../inputs/input3.txt') as f:
		lines = f.readlines()

	print('Result 1: {}'.format(getPowerConsumption(lines)))

	oxy = getLifeSupportRating(lines, mode=OXYGEN)
	co2 = getLifeSupportRating(lines, mode=CO2)
	print('Result 2: {}'.format(oxy * co2))
	